packages/core/src/sync-store/index.ts

```diff
@@ -1089,13 +1090,102 @@ export const createSyncStore = ({
   getRpcRequestResults: async ({ requests, chainId }) => {
     if (requests.length === 0) return [];
 
+    // Optimized fast path for high number of `requests` using a range of block numbers
+    // rather than querying each request individually.
+
+    const blockNumberRequests: (EIP1193Parameters & { blockNumber: number })[] =
+      [];
+    const nonBlockNumberRequests: EIP1193Parameters[] = [];
+    for (const request of requests) {
+      const blockNumber = extractBlockNumberParam(request);
+
+      // Note: "latest" is sorted into `nonBlockNumberRequests`
+      if (isHex(blockNumber)) {
+        blockNumberRequests.push({
+          ...request,
+          blockNumber: hexToNumber(blockNumber),
+        });
+      } else {
+        nonBlockNumberRequests.push(request);
+      }
+    }
+
     const requestHashes = requests.map((request) =>
       crypto
         .createHash("md5")
         .update(toLowerCase(JSON.stringify(orderObject(request))))
         .digest("hex"),
     );
 
```

Comment:
Maybe something like this would avoid an extra loop and recomputation of hashes:

```suggestion
    const blockNumberRequests: (EIP1193Parameters & { blockNumber: number })[] =
      [];
    const nonBlockNumberRequests: (EIP1193Parameters & { hash: string })[] = [];
    const requestHashes: string[] = [];
    for (const request of requests) {
    
      const hash =  crypto
        .createHash("md5")
        .update(toLowerCase(JSON.stringify(orderObject(request))))
        .digest("hex");
        
      const blockNumber = extractBlockNumberParam(request);

      // Note: "latest" is sorted into `nonBlockNumberRequests`
      if (isHex(blockNumber)) {
        blockNumberRequests.push({
          ...request,
          blockNumber: hexToNumber(blockNumber),
        });
      } else {
        nonBlockNumberRequests.push({...request, hash});
      }
    }

   requestHashes.push(hash);
```